File: YEAR 2/SEM 1/Advanced Programming Methods/LAB 4/src/Util/Stack/MyStack.java
package Util.Stack;

import java.util.Stack;

public class MyStack<E> implements MyIStack<E> {
    private Stack<E> stack;

    public MyStack() {
        this.stack = new Stack<>();
    }

    @Override
    public void push(E element) {
        this.stack.push(element);
    }

    @Override
    public E pop() throws StackException {
        if(this.isEmpty())
            throw new StackException("Cannot pop from empty stack!");

        return this.stack.pop();
    }

    @Override
    public E top() throws StackException {
        if(this.isEmpty())
            throw new StackException("Stack is empty!");

        return this.stack.peek();
    }

    @Override
    public boolean isEmpty() {
        return this.stack.isEmpty();
    }

    @Override
    public String toString() {
        StringBuilder stringBuilder = new StringBuilder();

        for (E element : stack) {
            stringBuilder.append(element);
            stringBuilder.append("\r\n");
        }

        return stringBuilder.toString();
    }
}

File: YEAR 2/SEM 1/Advanced Programming Methods/ToyLanguage/ToyLanguage/Util/Stack/MyStack.cs
using System.Collections.Generic;
using System.Text;

namespace ToyLanguage.Util.Stack
{
    class MyStack<T> : MyIStack<T>
    {
        private readonly Stack<T> stack = new Stack<T>();

        public bool IsEmpty => stack.Count == 0;

        public T Pop()
        {
            if (IsEmpty) throw new MyStackException("Pop from empty stack!");

            return stack.Pop();
        }

        public void Push(T item)
        {
            stack.Push(item);
        }

        public T Peek()
        {
            if (IsEmpty) throw new MyStackException("Top on empty stack!");

            return stack.Peek();
        }

        public override string ToString()
        {
            StringBuilder builder = new StringBuilder();
            var items = stack.ToArray();

            for(int i = 0; i < items.Length;i++)
            {
                builder.Append(items[i]);
                builder.Append(System.Environment.NewLine);
            }

            return builder.ToString();
        }
    }
}
